### PR TITLE
WIP: Remove gogo from the codebase

### PR DESCRIPTION
**Pull request by sdake**
GOGO adds very little performance value and adds alot of
complexity and misunderstanding (by developers). Remove it from
the code base.

The approach here is to permit unmarshal of gogo interfaces, but once
unmarhsled, all objects will become protobuf interfaces.

Related PR: https://github.com/istio/api/pull/1607

Please provide a description for what this PR is for.

And to help us figure out who should review this PR, please 
put an X in all the areas that this PR affects.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pilot/pkg/model/config.go [hunk 1]
```diff
@@ -127,14 +127,24 @@ type ConfigMeta struct {
...
+	Reset()
+	String() string
+	ProtoMessage()
+}
+
+// Config is a configuration unit consisting of the google.golang.org/protobtufype of configuration, the
+// key identifier that is unique per type, and the content represented as a protobuf message.
 type Config struct {
 	ConfigMeta
+	Spec Message
```
Comment (howardjohn) [inline pilot/pkg/model/config.go:141]: I am not actually sure this changes anything? Our `Message` interface is exactly the same as the previous proto.Message?

Comment (sdake) [reply pilot/pkg/model/config.go:141]: agree - I was just pushing my WIP in the event others wish to follow along as I flog the various repositories.

File: pkg/config/validation/validation.go [hunk 2]
```diff
@@ -848,9 +850,6 @@ func validateSidecarIngressPortAndBind(port *networking.Port, bind string) (errs
 }
 
 func validateTrafficPolicy(policy *networking.TrafficPolicy) error {
-	if policy == nil {
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:851]: should keep this?

File: pkg/config/validation/validation.go [hunk 3]
```diff
@@ -1024,46 +1019,18 @@ func ValidateProxyAddress(hostAddr string) error {
...
-	dur, err := types.DurationFromProto(pd)
-	if err != nil {
-		return err
-	}
-	if dur < time.Millisecond {
+	duration := pd.AsDuration()
+	if time.Duration.Milliseconds(duration) > 1 {
 		return errors.New("duration must be greater than 1ms")
 	}
-	if dur%time.Millisecond != 0 {
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:1051]: I think we are probably loosing this check?

File: pkg/config/validation/validation.go [hunk 4]
```diff
@@ -1083,32 +1050,32 @@ func ValidateParentAndDrain(drainTime, parentShutdown *types.Duration) (errs err
 		return
 	}
 
-	drainDuration, _ := types.DurationFromProto(drainTime)
-	parentShutdownDuration, _ := types.DurationFromProto(parentShutdown)
+	drainDuration := drainTime.GetSeconds()
+	parentShutdownDuration := parentShutdown.GetSeconds()
 
-	if drainDuration%time.Second != 0 {
+	if drainDuration < 1.0 {
```
Comment (howardjohn) [inline pkg/config/validation/validation.go:1056]: we are losing this check. Its about a user providing `1100ms` - we can only support second increments

File: pkg/config/validation/validation.go (hunk 2 above)
Comment (sdake) [reply pkg/config/validation/validation.go:851]: thanks. resolved.

File: pkg/config/validation/validation.go (hunk 3 above)
Comment (sdake) [reply pkg/config/validation/validation.go:1051]: time.Duration stores an int64 in nanoseconds: https://golang.org/src/time/time.go?s=20758:20777#L573 although my check on 1025 has the wrong conditional

Comment (howardjohn) [reply pkg/config/validation/validation.go:1051]: The problem is envoy does not - so if you specify some number of nanosecond precision it won't work

Comment (sdake) [reply pkg/config/validation/validation.go:1051]: I see, thanks for the clarity.

File: pkg/config/validation/validation.go (hunk 4 above)
Comment (sdake) [reply pkg/config/validation/validation.go:1056]: makes sense. I'll add a comment in this block as well - this is not really all that obvious.

**Comment by hzxuzhonghu** [comment]
Yesterday i have read a blog that gogo protobuf has better performance that golang

**Comment by sdake** [comment]
yes, although gogo is unmaintained. There is a longer thread here: https://github.com/istio/api/pull/1607#issuecomment-679468710

The request from T&R is to run the microbenchmarks for EDS - if the tradeoff makes sense then proceed.

Cheers,
-steve

**Comment by howardjohn** [comment]
The performance of protobuf marshalling for Istio APIs is not important at
all. For 99.99% of users, we literally never proto encode/decode them
(which is what gogo is fast at). We get it as JSON from the API server.

I am actually not sure if the json encoding performance is different, I am
pretty sure the code is identical. Even IF it is, even on absolutely
massive clusters the performance of marshalling Istio types is a trivial
fraction of total CPU usage.

Additionally, from initial tests I think our binary sizes will drop a fair
amount with the change (will test more once this PR is building fully).

tl;dr performance is "worse" in benchmarks but I doubt it will impact us.
We will do proper testing to verify.

**Comment by mandarjog** [comment]
I am closing this PR since there is no activity on it. Unless there is champion to remove gogo, it will not happen.
Feel free to re-open.
